File: sdk/aip/typescript/src/did.ts
/**
 * DID:KEY encoding/decoding (Multicodec + Base58btc)
 */

const ED25519_MULTICODEC = new Uint8Array([0xed, 0x01]);
const B58_ALPHABET = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz';

export function base58Encode(bytes: Uint8Array): string {
  const digits = [0];
  for (const byte of bytes) {
    let carry = byte;
    for (let j = 0; j < digits.length; j++) {
      carry += digits[j] << 8;
      digits[j] = carry % 58;
      carry = (carry / 58) | 0;
    }
    while (carry > 0) {
      digits.push(carry % 58);
      carry = (carry / 58) | 0;
    }
  }
  let output = '';
  for (const byte of bytes) {
    if (byte === 0) output += '1';
    else break;
  }
  for (let i = digits.length - 1; i >= 0; i--) {
    output += B58_ALPHABET[digits[i]];
  }
  return output;
}

export function base58Decode(str: string): Uint8Array {
  const bytes = [0];
  for (const char of str) {
    const index = B58_ALPHABET.indexOf(char);
    if (index < 0) throw new Error(`Invalid base58 character: ${char}`);
    let carry = index;
    for (let j = 0; j < bytes.length; j++) {
      carry += bytes[j] * 58;
      bytes[j] = carry & 0xff;
      carry >>= 8;
    }
    while (carry > 0) {
      bytes.push(carry & 0xff);
      carry >>= 8;
    }
  }
  for (const char of str) {
    if (char === '1') bytes.push(0);
    else break;
  }
  return new Uint8Array(bytes.reverse());
}

export function publicKeyToDIDKey(pubBytes: Uint8Array): string {
  if (pubBytes.length !== 32) throw new Error(`Expected 32 bytes, got ${pubBytes.length}`);
  const mc = new Uint8Array(34);
  mc.set(ED25519_MULTICODEC, 0);
  mc.set(pubBytes, 2);
  return `did:key:z${base58Encode(mc)}`;
}

export function didKeyToPublicKey(did: string): Uint8Array {
  if (!did.startsWith('did:key:z')) throw new Error('Invalid did:key format');
  const decoded = base58Decode(did.slice(9));
  if (decoded.length !== 34 || decoded[0] !== 0xed || decoded[1] !== 0x01) {
    throw new Error('Invalid did:key: wrong prefix or length');
  }
  return decoded.slice(2);
}

export function validateDIDKey(did: string): boolean {
  try {
    didKeyToPublicKey(did);
    return true;
  } catch {
    return false;
  }
}

File: sdk/aip/python/aip_sdk/did.py
"""DID:KEY encoding/decoding (Ed25519 Multicodec + Base58btc)."""

ED25519_MULTICODEC = b"\xed\x01"
B58_ALPHABET = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

def _b58encode(data: bytes) -> str:
    n = int.from_bytes(data, "big")
    result = []
    while n > 0:
        n, r = divmod(n, 58)
        result.append(B58_ALPHABET[r:r+1])
    for byte in data:
        if byte == 0: result.append(b"1")
        else: break
    return b"".join(reversed(result)).decode("ascii")

def _b58decode(s: str) -> bytes:
    n = 0
    for c in s:
        n = n * 58 + B58_ALPHABET.index(c.encode())
    # Count only leading '1' characters (not all)
    pad = 0
    for c in s:
        if c == "1": pad += 1
        else: break
    result = n.to_bytes((n.bit_length() + 7) // 8, "big") if n > 0 else b""
    return b"\x00" * pad + result

def public_key_to_did_key(pub_bytes: bytes) -> str:
    assert len(pub_bytes) == 32
    return "did:key:z" + _b58encode(ED25519_MULTICODEC + pub_bytes)

def did_key_to_public_key(did: str) -> bytes:
    if not did.startswith("did:key:z"): raise ValueError(f"Invalid did:key: {did[:20]}")
    decoded = _b58decode(did[9:])
    if len(decoded) != 34 or decoded[:2] != ED25519_MULTICODEC:
        raise ValueError("Invalid did:key: wrong prefix or length")
    return decoded[2:]

def validate_did_key(did: str) -> bool:
    try: did_key_to_public_key(did); return True
    except: return False
